hal/src/peripherals/adc/config.rs

```diff
@@ -0,0 +1,173 @@
+use atsamd_hal_macros::hal_cfg;
+
+#[hal_cfg(any("adc-d5x"))]
+use crate::pac::adc0;
+
+#[hal_cfg(any("adc-d21", "adc-d11"))]
+use crate::pac::adc as adc0;
+
+#[hal_cfg(any("adc-d21", "adc-d11"))]
+pub use adc0::ctrlb::Prescalerselect as AdcDivider;
+
+#[hal_cfg(any("adc-d5x"))]
+use adc0::ctrla::Prescalerselect as AdcDivider;
+
+pub use adc0::avgctrl::Samplenumselect as AdcSampleCount;
+
+pub use adc0::ctrlb::Resselselect as AdcResolution;
+pub use adc0::refctrl::Refselselect;
+
+/// Result accumulation strategy for the ADC
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum AdcAccumulation {
+    /// The ADC will read once and then the result is ready
+    Single,
+    /// The ADC will read [AdcSampleCount] samples, average them out
+    /// into a 16 bit wide value, and then the result is ready
+    Average(AdcSampleCount),
+    /// The ADC will read [AdcSampleCount] samples, sum them
+    /// into a 16 bit wide value, and then the result is ready
+    Summed(AdcSampleCount),
+}
+
+#[derive(Copy, Clone)]
+pub enum VrefSource {}
+
+/// # ADC configuration builder
+///
+/// Multiple factors can affect the ADCs overall sampling rate, and this
+/// structure allows for the configuring of the majority of factors that affect
+/// the sample rate of the ADC
+///
+/// To begin with, the ADC Clock is driven by the peripheral clock divided with
+/// a divider ([AdcDivider]).
+///
+/// Each sample is read by the ADC over
+/// [AdcSettingsBuilder::sample_clock_cycles] clock cycles, and then transmitted
+/// to the ADC register over [AdcSettingsBuilder::bit_width] clock cycles (1
+/// clock cycle per bit)
+///
+/// The ADC can also be configured to combine multiple simultaneous readings in
+/// either an average or summed mode (See [AdcAccumulation]), this also affects
+/// the overall sample rate of the ADC as the ADC has to do multiple
+/// samples before a result is ready.
+///
+/// Therefore, the overall formula for calculating Sample rate (SPS) can be
+/// calculated like so:
+///
+/// ## For single sample
+/// ```
+/// SPS = (GCLK_ADC / clk_divider) / (sample_clock_cycles + bit_width)
+/// ```
+/// ## For multiple samples (Averaging or Summed)
+/// ```
+/// SPS = (GCLK_ADC / clk_divider) / (n * (sample_clock_cycles + bit_width))
+/// ```
+#[derive(Copy, Clone)]
```

Comment:
I don't think it's appropriate for the `Config` type to be `Copy`.

Reply:
Copy just makes it a little easier if you want to use the same ADC Settings for both ADC0 and ADC1 (in the case of sam5x)